### ISSUE TITLE
theano.tensor.switch produces nan gradients

### ISSUE THREAD

**Issue by Ignotus** [Theano/Theano#5333]
Hello again,

I have been faced with nan gradients occurred when using with T.switch. Below is a snippet to reproduce results:

```python
import numpy as np
import theano
import theano.tensor as T
theano.config.floatX = 'float32'

def ef(w):
    t = T.sum(w ** 2, axis=1)

    # Fancy limit
    f = T.sin(t) / t

    R = T.switch(t < 1e-12, 1., f)
    return R

w = T.fmatrix()
exp = ef(w)

loss = T.sum(exp)
grad = T.grad(loss, w)

f = theano.function([w], [exp, grad, loss])


array = np.array([[1, 2, 1],
                  [0, 0, 0]], dtype=np.float32)

for i in range(2):
    e, g, l = f(array)
    # print(e.shape, g.shape, l.shape)
    print(e, l, g)
    array[:, :3] -= 0.01 * g
``` 

Best Regards,

Minh

**Comment by lamblin** [comment]
I just tried with ad1310c88, and it worked, producing the following:
```(array([-0.04656925,  1.        ], dtype=float32), array(0.9534307718276978, dtype=float32), array([[ 0.33557984,  0.67115968,  0.33557984],
       [ 0.        ,  0.        ,  0.        ]], dtype=float32))
(array([-0.05332064,  1.        ], dtype=float32), array(0.9466793537139893, dtype=float32), array([[ 0.33495238,  0.66990477,  0.33495238],
       [ 0.        ,  0.        ,  0.        ]], dtype=float32))
```
Do you have anything in your Theano config that could prevent optimizations?
Which exact commit are you using?

**Comment by Ignotus** [comment]
I have the following config:
```
[global]
floatX = float32
device = gpu0
warn_float64 = raise

[lib]
cnmem=.85
```

May warn_float64 cause the problem?

**Comment by lamblin** [comment]
I just tried with `warn_float64=raise`, and again, it is working for me.
Which version (commit number) of Theano are you trying it with?
Did you try with the latest dev branch?
